### PR TITLE
More accurate estimation of the size of a RAID

### PR DESCRIPTION
**Pull request by mwhudson**
I found the code in mdadm that decides how much of a device is usable
for data and copied it. It's a bit hairy!

For https://bugs.launchpad.net/subiquity/+bug/1816777 and many similar
bugs.

### CODE REVIEW

**Comment by xnox** [comment]
i guess i want VM tests for all raid levels at this point.

File: subiquity/models/filesystem.py [hunk 1]
```diff
@@ -218,24 +218,38 @@ def dehumanize_size(size):
...
+    # touch ridiculous. What follows is a translation of the code at:
+    # https://git.kernel.org/pub/scm/utils/mdadm/mdadm.git/tree/super1.c?h=mdadm-4.1&id=20e8fe52e7190b3ffda127566852eac2eb7fa1f7#n2770
+    # (note that that calculation is in terms of 512-byte sectors and
+    # this one is in bytes).
+    #
+    # This makes assumptions about the defaults mdadm uses but mostly
+    # that the default metadata version is 1.2, and other formats use
+    # less space.
+    min_size = min(dev.size for dev in devices)
+    bmspace = 128*1024
```
Comment (raharper) [inline subiquity/models/filesystem.py]: I'd expand the variable names here, since it's not kernel code.
bitmap_space, and summarize the decision, we're assuming
most devices are going to be > 200GiB so we select the 128k bitmap_space

File: subiquity/models/filesystem.py [hunk 2]
```diff
@@ -218,24 +218,38 @@ def dehumanize_size(size):
...
+    headroom = 128*1024*1024
```
Comment (raharper) [inline subiquity/models/filesystem.py]: What does headroom correspond to?

File: subiquity/models/filesystem.py [hunk 3]
```diff
@@ -218,24 +218,38 @@ def dehumanize_size(size):
...
+    while (headroom << 10) > min_size and headroom > 2*1024*1024:
+        headroom >>= 1
+    # mdadm's Create() can round things a little more so, to be
+    # pessimistic, assume another megabyte gets wasted somewhere.
+    data_offset = align_up(12*1024 + bmspace + headroom) + 1024*1024
```
Comment (raharper) [inline subiquity/models/filesystem.py]: I'd be surprised if this ends up being larger than the 8MB overhead you already had.

File: subiquity/models/tests/test_filesystem.py [hunk 4]
```diff
@@ -105,6 +106,12 @@ def test_errors(self):
                 self.assertEqual(expected_error, actual_error)
 
 
+class TestRoundRaidSize(unittest.TestCase):
+
+    def test_lp1816777(self):
+        self.assertLessEqual(round_raid_size(500107862016), 499972571136)
```
Comment (raharper) [inline subiquity/models/tests/test_filesystem.py]: 4.6 TiB drive has 130 MB overhead.  OK.  It would be interesting to see the mdadm --examine superblock dump to match up where the space went (bitmap, alignment, etc).

**Comment by raharper** [comment]
> i guess i want VM tests for all raid levels at this point.

We don't need VM tests so much as cycling values through
the calculation, then running mdadm create on a block device
and using mdadm examine to verify the overhead and that
we can successfully create it.

File: subiquity/models/filesystem.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/models/filesystem.py]: I think there's some value in following the variable names from the mdadm source (maybe I should paste that in a comment) but you're right about where the value for bmspace comes from.

File: subiquity/models/filesystem.py (hunk 2 above)
Comment (mwhudson) [reply subiquity/models/filesystem.py]: No idea really! The mdadm source says "allow for some headroom for reshape".

File: subiquity/models/filesystem.py (hunk 3 above)
Comment (mwhudson) [reply subiquity/models/filesystem.py]: Get ready to be surprised!

File: subiquity/models/tests/test_filesystem.py (hunk 4 above)
Comment (mwhudson) [reply subiquity/models/tests/test_filesystem.py]: Sadly none of our bug reporters we quite as thorough as to include that :)

**Comment by mwhudson** [comment]
> > i guess i want VM tests for all raid levels at this point.
> 
> We don't need VM tests so much as cycling values through
> the calculation, then running mdadm create on a block device
> and using mdadm examine to verify the overhead and that
> we can successfully create it.

Yeah, indeed. I don't think there's any way to write this test without requiring root so I don't know if we can make anything like this a unit test but it's probably still worth writing these tests...

File: scripts/raid-size-tests.py [hunk 5]
```diff
@@ -89,36 +92,41 @@ class FakeDev:
...
+        if len(set(sizes)) == 1:
+            sz = '[{}]*{}'.format(humanize_size(sizes[0]), len(sizes))
+        else:
+            sz = str([humanize_size(s) for s in sizes])
+        print("level {} sizes {} -> calc_size {} real_size {}".format(
+            level, sz , calc_size, real_size), end=' ')
+        if calc_size > real_size:
+            print("BAAAAAAAAAAAD", real_size - calc_size)
+        else:
+            print("OK by", real_size - calc_size)
```
Comment (raharper) [inline scripts/raid-size-tests.py]: print("RAID overhead %s bytes" % real_size - calc_size)

Comment (mwhudson) [reply scripts/raid-size-tests.py]: That's not the RAID overhead though, it's the difference between our estimate of the size and the real size. I'll make this a bit clearer.

**Comment by mwhudson** [comment]
I ran some tests and found that my code wasn't right when the sizes of the devices varied. I've updated it to match mdadm's behaviour in that regard and now raid-size-tests.py reports subiquity's estimates as byte accurate (apart from raid10 with odd numbers of devices). It's a bit more involved of a change now, so I'll let it sit for a while before merging...

**Comment by xnox** [comment]
Mdadm doesn't expose this at all does it? Can we like execute mdadm, or its library to give us the numbers? I.e. dryrun on the device?!

**Comment by mwhudson** [comment]
Nope, mdadm doesn't do anything that useful. subiquity *could* do the "loopback devices in a tmpfs" game my test script does, but that sounds insane (it's also quite slow).

**Comment by mwhudson** [comment]
Hmm that's not a great filename.

**Comment by mwhudson** [comment]
And this was the wrong PR of course.

File: scripts/raid-size-tests.py [hunk 6]
```diff
@@ -107,8 +115,14 @@ def verify_size_ok(level, sizes):
...
             print("BAAAAAAAAAAAD", real_size - calc_size)
+            if os.environ.get('DEBUG'):
+                print(raid)
+                input('waiting: ')
+        elif calc_size == real_size:
+            print("exactly right!")
+            r = True
         else:
-            print("OK by", real_size - calc_size)
+            print("underestimated size by", real_size - calc_size)
```
Comment (raharper) [inline scripts/raid-size-tests.py]: subiquity wasted space?

File: subiquity/controllers/filesystem.py [hunk 7]
```diff
@@ -419,13 +419,13 @@ def delete_raid(self, raid):
         self.model.remove_raid(raid)
 
     def create_volgroup(self, spec):
-        devices = set()
+        devices = []
```
Comment (raharper) [inline subiquity/controllers/filesystem.py]: One can't add the same disk twice, so set() seems in order here.  What's the reason for the change?

File: subiquity/models/filesystem.py [hunk 8]
```diff
@@ -218,36 +218,63 @@ def dehumanize_size(size):
...
+# We make some assumptions about the defaults mdadm uses but mostly that the
+# default metadata version is 1.2, and other formats use less space.
+#
+# Note that data_offset is computed for the first disk mdadm examines and then
+# used for all devices, so the order matters! (Well, if the size of the
+# devices vary, which is not normal but also not something we prevent).
+#
+# All this is tested against reality in ./scripts/get-raid-sizes.py
+def calculate_data_offset(devsize):
+    devsize >>= 9  # convert to sectors
```
Comment (raharper) [inline subiquity/models/filesystem.py]: I think we could add a comment here to the effect that we're using sectors to make it easier to match the kernel mdadm algorithm and we convert back to bytes.

I'd also suggest renaming this function to 'calculate_data_offset_bytes'

File: scripts/raid-size-tests.py (hunk 6 above)
Comment (mwhudson) [reply scripts/raid-size-tests.py]: OK.

File: subiquity/controllers/filesystem.py (hunk 7 above)
Comment (mwhudson) [reply subiquity/controllers/filesystem.py]: Well the discovery I made yesterday that the order matters to mdadm. But I agree this is conceptually bogus. Probably the better fix is to pass them around as a set and sort them (consistently but arbitrarily) in get_raid_sizes and when rendering the config.

That will let me make this change much smaller, which is nice...

File: subiquity/models/filesystem.py (hunk 8 above)
Comment (mwhudson) [reply subiquity/models/filesystem.py]: Fair on both counts.
